File: topos/evaluation/policies/secure.py
"""
Φ_SECURE: Policy translator for the SECURE generator.
-----------------------------------------------------

Maps CPG-based security observations into a
:class:`~topos.evaluation.policies.base.ScoredDecision`.
``achieved`` requires zero dangerous calls and zero taint flows;
``score`` is ``min(per-metric qualities)`` for reporting only.

Quality functions:
    danger_quality = exp(-dangerous_calls / danger_scale)
    taint_quality  = exp(-taint_flows / taint_scale)

The SECURE badge is achieved if and only if there are zero dangerous calls
and zero taint flows (strict security). Gate comparisons and interpretation
prose live in :mod:`topos.evaluation.policies.gates`; thresholds in
:mod:`topos.evaluation.policies.calibration`.
"""

from __future__ import annotations

from math import exp

from topos.evaluation.policies.base import (
    Priority,
    ScoredDecision,
)
from topos.evaluation.policies.calibration import SECURE
from topos.evaluation.policies.gates import evaluate_gates


def score_secure(
    dangerous_calls: float | None = None,
    taint_flows: float | None = None,
    priority: Priority = Priority.SECURE,
    threshold: float | None = None,
) -> ScoredDecision:
    """
    Φ_SECURE — score the SECURE generator from CPG observations.

    Args:
        dangerous_calls: Count of reachable dangerous-API call sites.
        taint_flows:     Count of source→sink data-flow paths.
        priority:        Retained for API compatibility; not read by this Φᵢ.
        threshold:       Retained for API compatibility; not read by this Φᵢ.

    Returns:
        A ScoredDecision; ``achieved`` is the truth value of the SECURE
        generator for this program.
    """
    metrics = {
        key: value
        for key, value in {
            "cpg.dangerous_calls": dangerous_calls,
            "cpg.taint_flows": taint_flows,
        }.items()
        if value is not None
    }
    results = evaluate_gates(metrics, pillar="secure")
    if not results:
        # If no metrics are provided, we vacuously satisfy SECURE.
        return ScoredDecision(score=1.0, achieved=True, interpretation={})

    # Score shaping (reporting only): exponential decay stays local to Φ_SECURE.
    scale = {
        "cpg.dangerous_calls": SECURE.danger_scale,
        "cpg.taint_flows": SECURE.taint_scale,
    }
    qualities = [exp(-max(r.value, 0.0) / scale[r.spec.metric]) for r in results]

    return ScoredDecision(
        # The combined score is the minimum of the individual qualities
        # (conservative AND).
        score=min(qualities),
        achieved=all(r.passed for r in results),
        interpretation={r.spec.metric: r.interpretation for r in results},
    )

File: crates/topos-core/src/evaluation/policies/secure.rs
//! `Φ_SECURE`: policy translator for the SECURE generator.
//!
//! Maps CPG-based security observations into a [`ScoredDecision`].
//! `achieved` requires zero dangerous calls and zero taint flows;
//! `score` is `min(per-metric qualities)` for reporting only.
//!
//! Quality functions:
//! - `danger_quality = exp(-dangerous_calls / danger_scale)`
//! - `taint_quality  = exp(-taint_flows / taint_scale)`
//!
//! The SECURE badge is achieved if and only if there are zero dangerous
//! calls and zero taint flows (strict security). Gate comparisons and
//! interpretation prose live in [`super::gates`]; thresholds in
//! [`super::calibration`].

use std::collections::HashMap;

use super::base::ScoredDecision;
use super::calibration::SECURE;
use super::gates::evaluate_gates;

/// `Φ_SECURE` — score the SECURE generator from CPG observations.
pub fn score_secure(dangerous_calls: f64, taint_flows: f64) -> ScoredDecision {
    let metrics = HashMap::from([
        ("cpg.dangerous_calls".to_string(), dangerous_calls),
        ("cpg.taint_flows".to_string(), taint_flows),
    ]);

    let results = evaluate_gates(&metrics, Some("secure"), false);
    if results.is_empty() {
        // If no metrics are provided, we vacuously satisfy SECURE.
        return ScoredDecision {
            score: 1.0,
            achieved: true,
            interpretation: HashMap::new(),
        };
    }

    // Score shaping (reporting only): exponential decay stays local to Φ_SECURE.
    let qualities: Vec<f64> = results
        .iter()
        .map(|r| {
            let scale = if r.spec.metric == "cpg.dangerous_calls" {
                SECURE.danger_scale
            } else {
                SECURE.taint_scale
            };
            (-r.value.max(0.0) / scale).exp()
        })
        .collect();

    ScoredDecision {
        score: qualities.into_iter().fold(f64::INFINITY, f64::min),
        achieved: results.iter().all(|r| r.passed()),
        interpretation: results
            .iter()
            .map(|r| (r.spec.metric.to_string(), r.interpretation()))
            .collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clean_code_scores_one() {
        let result = score_secure(0.0, 0.0);
        assert_eq!(result.score, 1.0);
        assert!(result.achieved);
    }

    #[test]
    fn dangerous_code_scores_low_and_fails() {
        let result = score_secure(20.0, 20.0);
        assert!(result.score < 0.1);
        assert!(!result.achieved);
    }

    #[test]
    fn independent_thresholds_each_fail_alone() {
        assert!(score_secure(0.0, 0.0).achieved);
        assert!(!score_secure(1.0, 0.0).achieved);
        assert!(!score_secure(0.0, 1.0).achieved);
    }
}
